### PR TITLE
[FLINK-20949][table-planner-blink] Separate the implementation of sink nodes

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Separate the implementation of sink nodes, including StreamExecSink, BatchExecSink, StreamExecLegacySink, BatchExecLegacySink*

## Brief change log

  - *Introduce StreamPhysicalSink, and make StreamExecSink only extended from ExecNode*
  - *Introduce BatchPhysicalSink, and make BatchExecSink only extended from ExecNode*
  - *Introduce StreamPhysicalLegacySink, and make StreamExecLegacySink only extended from ExecNode*
  - *Introduce BatchPhysicalLegacySink, and make BatchExecLegacySink only extended from ExecNode*
  - *BatchCommonSubGraphBasedOptimizer should also consider Sink node*

## Verifying this change

This change is a refactoring rework covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
Thanks for the review @wenlong88 , I have updated the pr based on comments

**Review by wenlong88 — COMMENTED** [review]
thanks for the update, the overall pr LGTM, just one minor suggestion: maybe we can share more code between BatchExecLegacySink and StreamExecLegacySink, since most of the transforming logic are the same.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLegacySink.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        }
+    }
+
+    /**
+     * Translates {@link TableSink} into a {@link Transformation}.
+     *
+     * @param withChangeFlag Set to true to emit records with change flags.
+     * @return The {@link Transformation} that corresponds to the translated {@link TableSink}.
+     */
+    protected abstract Transformation<T> translateToTransformation(
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLegacySink.java:145]: how about we just make the abstract interface which just returns the converted inputRowType, like: createActualSinkInputType. the name and logic is a bit of confusing here.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLegacySink.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            PlannerBase planner, boolean withChangeFlag);
+
+    @SuppressWarnings("unchecked")
+    protected Transformation<T> createSinkTransformation(
```
Comment (wenlong88) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecLegacySink.java:149]: rename to: createSinkInputTransformation?

**Review by wenlong88 — APPROVED** [review]

**Pull request closed without merging**